src/main/java/org/skife/jdbi/v2/unstable/BindIn.java

```diff
@@ -25,82 +25,72 @@
 import java.lang.annotation.RetentionPolicy;
 import java.lang.reflect.Method;
-import java.sql.SQLException;
 import java.util.ArrayList;
-import java.util.Collection;
 import java.util.Iterator;
 import java.util.List;
 
+/**
+ * Binds an Iterable or array/varargs to a list of arguments (e.g. for WHERE X IN (...) query statements)
+ */
 @Retention(RetentionPolicy.RUNTIME)
 @SqlStatementCustomizingAnnotation(BindIn.CustomizerFactory.class)
 @BindingAnnotation(BindIn.BindingFactory.class)
-public @interface BindIn
-{
+public @interface BindIn {
     String value();
 
-    final class CustomizerFactory implements SqlStatementCustomizerFactory
-    {
+    final class CustomizerFactory implements SqlStatementCustomizerFactory {
 
         @Override
-        public SqlStatementCustomizer createForMethod(Annotation annotation, Class sqlObjectType, Method method)
-        {
+        public SqlStatementCustomizer createForMethod(final Annotation annotation, final Class sqlObjectType, final Method method) {
             throw new UnsupportedOperationException("Not supported on method!");
         }
 
         @Override
-        public SqlStatementCustomizer createForType(Annotation annotation, Class sqlObjectType)
-        {
+        public SqlStatementCustomizer createForType(final Annotation annotation, final Class sqlObjectType) {
             throw new UnsupportedOperationException("Not supported on type");
         }
 
         @Override
-        public SqlStatementCustomizer createForParameter(Annotation annotation,
-                                                         Class sqlObjectType,
-                                                         Method method,
-                                                         Object arg)
-        {
-            Collection<?> coll = (Collection<?>) arg;
-            BindIn in = (BindIn) annotation;
-            final String key = in.value();
-            final List<String> ids = new ArrayList<String>();
-            for (int idx = 0; idx < coll.size(); idx++) {
-                ids.add("__" + key + "_" + idx);
+        public SqlStatementCustomizer createForParameter(final Annotation annotation, final Class sqlObjectType, final Method method, final Object arg) {
+            final BindIn bindIn = (BindIn) annotation;
+            final String key = bindIn.value();
+            final int size = Objects.arraySize(arg);
+
+            final List<String> ids = new ArrayList<String>(size);
+            for (int i = 0; i < size; i++) {
+                ids.add("__" + key + "_" + i);
             }
 
-            StringBuilder names = new StringBuilder();
-            for (Iterator<String> i = ids.iterator(); i.hasNext();) {
-                names.append(":").append(i.next());
-                if (i.hasNext()) {
+            final StringBuilder names = new StringBuilder();
+            for (final Iterator<String> it = ids.iterator(); it.hasNext(); ) {
+                names.append(":").append(it.next());
```

Comment:
This is sort of orthogonal to this PR, but it would create less objects if the previous for-loop were collapsed into this one:

    for (int i = 0; i < size; i++) {
      if (id > 0) {
        names.append(",");
      }
      names.append(":__").append(key).append('_').append(i);
    }